### PR TITLE
Speedup of canPassThrough() method

### PR DESCRIPTION
**Pull request by JackRainy**
I have been playing with JProfiler (amasing tool, btw! Unfortunately, I cannot afford it for myself) and found out that `canPassThrough()` method - the most frequent called method can be optimized and speed up.

Preliminary investigation:
`canPassThrough()` method calls `canEnterTiles()` which calls `isAtWarWith()` and it takes almost 50% of its own time. Another 49% of the time is spent for HashMap `contains` checks of the `uniques`.

I did refactoring of those two methods: `canEnterTiles()` does not call `isAtWarWith()` anymore, but uses its checks directly. Also, `isAtWarWith()` does not get diplomacy manager twice. That's has given quite good throughput already.

Furthermore, I removed one more redundant check `isEmpty`: method `first()` does a check `isEmpty` inside, so no need to call `isNotEmpty` before it.

And lastly, I used the same trick as we use in other places - caching of the MapUnit uniques as boolean flags - it also gave some portion of speed boost.

Results:
| Test set | Before | After | Boost |
|--------|-------|--------|-------|
|  Test set 1 | 15-16 sec | 11-13 sec | **19-27%** |
|  Test set 2 | 12.5 sec | 10-12 sec | **4-20%** |

Summing up: `canPassThrough()` is almost **20% faster** now!

And of course, I did 100% unit test coverage for the method before introducing any changes.

### CODE REVIEW

**Comment by yairm210** [comment]
Amazing!!!
I cannot tell you how much time I have spent on improving the performance of the basic movement functions, it's great to see that there's always room for more improvement ❤️
